src/bits32/paging.rs

```diff
@@ -0,0 +1,1527 @@
+}
+
+impl fmt::Octal for VAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::UpperHex for VAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl fmt::Pointer for VAddr {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use core::fmt::LowerHex;
+        self.0.fmt(f)
+    }
+}
+
+#[allow(clippy::clippy::derive_hash_xor_eq)]
+impl Hash for VAddr {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.0.hash(state);
+    }
+}
+
+/// Log2 of base page size (12 bits).
+pub const BASE_PAGE_SHIFT: usize = 12;
+
+/// Size of a base page (4 KiB)
+pub const BASE_PAGE_SIZE: usize = 4096;
+
+/// Size of a large page (2 MiB)
+pub const LARGE_PAGE_SIZE: usize = 1024 * 1024 * 2;
+
+/// Size of a huge page (1 GiB)
+pub const HUGE_PAGE_SIZE: usize = 1024 * 1024 * 1024;
+
+/// Size of a cache-line
+pub const CACHE_LINE_SIZE: usize = 64;
+
+/// A type wrapping a base page with a 4 KiB buffer.
+pub struct Page([u8; BASE_PAGE_SIZE]);
+
+/// A type wrapping a large page with a 2 MiB buffer.
+pub struct LargePage([u8; LARGE_PAGE_SIZE]);
+/// A type wrapping a huge page with a 1 GiB buffer.
+pub struct HugePage([u8; HUGE_PAGE_SIZE]);
+
+/// Mask to find the physical address of an entry in a page-table.
+const ADDRESS_MASK: u32 = !0xfff;
+
+/// Page tables have 512 = 4096 / 32 entries.
+pub const PAGE_SIZE_ENTRIES: usize = 1024;
+
+/// A page directory.
+pub type PD = [PDEntry; PAGE_SIZE_ENTRIES];
+
+/// A page table.
+pub type PT = [PTEntry; PAGE_SIZE_ENTRIES];
+
+/// Given virtual address calculate corresponding entry in PD.
+#[inline]
+pub fn pd_index(addr: VAddr) -> usize {
+    ((addr >> 22usize) & 0b1111111111) as usize
+}
+
+/// Given virtual address calculate corresponding entry in PT.
+#[inline]
+pub fn pt_index(addr: VAddr) -> usize {
+    ((addr >> 12usize) & 0b1111111111) as usize
+}
+
+bitflags! {
+    /// PD configuration bits description.
+    #[repr(transparent)]
+    pub struct PDFlags: u32 {
+        /// Present; must be 1 to map a 2-MByte page or reference a page table.
+        const P       = bit!(0);
+        /// Read/write; if 0, writes may not be allowed to the 2-MByte region controlled by this entry
+        const RW      = bit!(1);
+        /// User/supervisor; user-mode accesses are not allowed to the 2-MByte region controlled by this entry.
+        const US      = bit!(2);
+        /// Page-level write-through.
+        const PWT     = bit!(3);
+        /// Page-level cache disable.
+        const PCD     = bit!(4);
+        /// Accessed; if PS set indicates whether software has accessed the 2-MByte page
+        /// else indicates whether this entry has been used for linear-address translation
+        const A       = bit!(5);
+        /// Dirty; if PS indicates whether software has written to the 2-MByte page referenced by this entry.
+        /// else ignored.
+        const D       = bit!(6);
```

Comment:
AFAIK this one doesn't exist (must be zero) on 32bit?

Reply:
It seems to exist when it references 4-MiB page. (Section 4.3 of Intel Developer's Manual)